cpp/src/arrow/compute/api_scalar.h

```diff
@@ -49,10 +49,66 @@ class ARROW_EXPORT ElementWiseAggregateOptions : public FunctionOptions {
   explicit ElementWiseAggregateOptions(bool skip_nulls = true);
   constexpr static char const kTypeName[] = "ElementWiseAggregateOptions";
   static ElementWiseAggregateOptions Defaults() { return ElementWiseAggregateOptions{}; }
-
   bool skip_nulls;
 };
 
+/// Rounding and tie-breaking modes for round compute functions.
+/// Additional details and examples are provided in compute.rst.
+enum class RoundMode : int8_t {
+  /// Round to nearest integer less than or equal in magnitude (aka "floor")
+  DOWN,
+  /// Round to nearest integer greater than or equal in magnitude (aka "ceil")
+  UP,
+  /// Get the integral part without fractional digits (aka "trunc")
+  TOWARDS_ZERO,
+  /// Round negative values with DOWN rule and positive values with UP rule
+  TOWARDS_INFINITY,
+  /// Round ties with DOWN rule
+  HALF_DOWN,
+  /// Round ties with UP rule
+  HALF_UP,
+  /// Round ties with TOWARDS_ZERO rule
+  HALF_TOWARDS_ZERO,
+  /// Round ties with TOWARDS_INFINITY rule
+  HALF_TOWARDS_INFINITY,
+  /// Round ties to nearest even integer
+  HALF_TO_EVEN,
+  /// Round ties to nearest odd integer
+  HALF_TO_ODD,
+};
+
+static constexpr double kDefaultAbsoluteTolerance = 1E-5;
+
+class ARROW_EXPORT RoundOptions : public FunctionOptions {
+ public:
+  explicit RoundOptions(int64_t ndigits = 0,
+                        RoundMode round_mode = RoundMode::HALF_TO_EVEN,
+                        double abs_tol = kDefaultAbsoluteTolerance);
+  constexpr static char const kTypeName[] = "RoundOptions";
+  static RoundOptions Defaults() { return RoundOptions(); }
+  /// Rounding precision (number of digits to round to).
+  int64_t ndigits;
+  /// Rounding and tie-breaking mode
+  RoundMode round_mode;
+  /// Absolute tolerance for approximating values as integers and mid-point decimals
+  double abs_tol;
```

Comment:
Is it really desirable to implement this? If I have "1.500000000001", will it really be considered equal to 1.5?

Reply:
And in case an optional tolerance is actually useful, I would really make it 0 by default.

Reply:
With the current tolerance, 1.500000000001 will be equal to 1.5. *The tolerance is to be able to control equality approximations since floating point numbers are not equally spaced.* Minor floating point errors are introduced when the value is scaled during the rounding process. So if you have very small numbers (e.g. 0.00000034, 0.0000003401) you can set tolerance accordingly so that they are not equal, but if you have very large numbers (e.g., 10000.000001, 10000.000000) you may want to consider them equal. In reality, this parameter should be use in the general case, but I did found that Numpy's rounding also has its corner cases, so this is a best effort approach. Setting it to 0 will allow some of those corner cases, so what about a smaller tolerance, 1e-8? If not, that is fine, 0 it is to mimic the behavior of other libraries as well.

Reply:
If NumPy doesn't support it, is it super useful to implement at all?

Reply:
I think 0 as a default value is best (especially as `ndigits` is 0 by default as well), it will match the common expectation.

Reply:
Another option is to have it internally for comparing floating-point numbers but not expose it to the client code as an option.

Reply:
Ok, I am removing the absolute tolerance parameter.

Reply:
Sound good. If possible ask for it, we can put it back later.